### ISSUE TITLE
Allow Scenes to specify where children of instantiations get placed in hierarchy (Shadow DOM)

### ISSUE THREAD

**Issue by dmchurch** [godotengine/godot-proposals#8184]
### Describe the project you are working on

A 2D game using a Control-based UI with automatic layout nodes.

### Describe the problem or limitation you are having in your project

I can't place children of complex Scenes properly without turning on Editable Children.

![image](https://github.com/godotengine/godot-proposals/assets/605817/6edeb4c7-9157-4bfe-9783-e72e4d0ea4c9)
Here you can see the complex Scene that defines my "Main Console" UI (or at least, one possible "main console"). The actual game content should go in the "CenterArea" container, selected.

![image](https://github.com/godotengine/godot-proposals/assets/605817/313965f1-9c2e-4f95-bad8-0c0b20346dd0)
However, when I instantiate a MainConsole and add a child node, that child is placed as a child of the MainConsole's root element. I could turn on "Editable Children" and place the child node manually, but then I couldn't switch out my console or make changes to it without having to fix every place I've already instantiated it.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allowing Scenes themselves to specify where children are placed in the hierarchy improves encapsulation, as it allows "what nodes are children of the root" to become solely an implementation detail of the scene. In this case, the MainConsole scene specifies that child nodes should be placed under the "CenterArea" node, and the "main" scene doesn't have to have Editable Children expanded.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This is, in effect, an Editor-specific feature to add parity to something already possible at runtime. If I need to instantiate a scene at runtime and then add children under a specific child node, I'd export the relative `NodePath` to the child node in question and have the instantiating code place children there.

As such, this will be an editor annotation on an exported NodePath variable. Unless there are suggestions otherwise, I'd use the following GDscript syntax:

```gdscript
@export_default_child_path var default_child_path : NodePath
```

That would cause "Default Child Path" to show up as a property in the Inspector when editing the child scene. When instantiating a parent scene, a valid, non-null @export_default_child_path export present on the instantiated child's root causes children to get placed under that node rather than under the root node.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I haven't been able to work around this issue in script, and it doesn't seem that there is currently any way to do this at the engine level. Once this is implemented, it is entirely opt-in as it only has any effect with a properly-annotated property.

### Is there a reason why this should be core and not an add-on in the asset library?

It can't be done in script. It affects how the engine hydrates a `.tscn` file, as

```
[node name="ActiveArea" type="Control" parent="MainConsole"]
```

will need to add ActiveArea as a node under a different parent than MainConsole, but the Editor will still need to show it as parented under MainConsole when "Editable Children" is unchecked. (When it is checked, the nodes will appear at their actual position in the hierarchy, but if it is again unchecked the added children will return to their original hierarchy position.)

**Comment by dmchurch** [comment]
I've written a first-effort PR in https://github.com/godotengine/godot/pull/83635, though I don't expect it's ready for merge yet.

The current PR contains the following changes:
1. Adds a GDscript annotation `@export_default_child_path`, which maps to the internal `PROPERTY_HINT_DEFAULT_SCENE_CHILDREN_PATH`
2. When the root node in a scene has `@export_default_child_path` on a `NodePath` property, it saves the name of that property in the `[node]` header line for the root node of the .tscn, as in the following:
```ini
[node name="MainConsole" type="VBoxContainer" default_child_path="child_container"]
child_container = NodePath("MainArea/CenterArea")
```
3. When a PackedScene is loaded from a .tscn file, if the root node has a `default_child_path`, that name is stored in the `SceneState` data structure. Note that there is a little bit of disconnect here! the `default_child_path` attribute is saved with the _node_, but in the internal representation it's stored with the _scene_. I think it's the right decision on both counts, since I didn't want to add to the Node::Data structure without cause and it makes sense to put the property name on the `[node]` that owns the property, but I wanted to call that out.
4. When a PackedScene is instantiated, if that scene contains one or more instantiations of child scenes that have `default_child_path` specified (the presence or absence of `default_child_path` on the main scene has no impact), it will store a mapping from the instantiated sub-scene's root to the node specified by the property named as `default_child_path`, assuming that path is valid. When later nodes in that file are instantiated, if the specified parent is one of these sub-scene roots, it will instead be added as a child at the default child path.
5. In the Scene Tree, any descendants of a non-editable scene instance that _are not_ owned by that scene will be displayed as though they were direct children of the instance root. Checking "Editable Children" will show these non-owned descendants in their actual position in the hierarchy, and unchecking it will return them to their original position.
6. If the user attempts to reparent a node to an instance root and that instance is of a scene with `default_child_path` specified, that node will instead reparent to the default child path, UNLESS the instance is editable AND the user dropped the nodes at a specific child position under the instance root.
7. When a scene is saved, if it contains non-editable scene instances that have `default_child_path` specified, children of that path that aren't owned by the subscene are written to the .tscn file as though they were children of the non-editable instance root.

Pictures are worth a kiloword, though, so here's a definition of a scene using the new annotation:
![image](https://github.com/godotengine/godot-proposals/assets/605817/ff2949d1-5411-4f1f-92a0-a1e3f80bc439)

This scene saves as follows (abbreviated):
```ini
[gd_scene load_steps=4 format=3 uid="uid://cfgv7xik5c3u5"]

[sub_resource type="GDScript" id="GDScript_xm4y2"]
resource_name = "MainConsole"
script/source = "extends VBoxContainer

@export var clock : Clock
@export var transmission_state: Label
@export_default_child_path var child_container : NodePath
"

[node name="MainConsole" type="VBoxContainer" node_paths=PackedStringArray("clock", "transmission_state") default_child_path="child_container"]
script = SubResource("GDScript_xm4y2")
clock = NodePath("HeaderBG/Header/Clock")
transmission_state = NodePath("HeaderBG/Header/TransmissionState")
child_container = NodePath("MainArea/CenterArea")
```

This scene references the previous "MainConsole" scene and has a control that appears as a direct child of the MainConsole:
![image](https://github.com/godotengine/godot-proposals/assets/605817/77b36933-dafd-4167-a3d2-eeffd6359ff2)

Activating "Editable Children" for the MainConsole shows the ActiveArea descendant at its actual position in the tree:
![image](https://github.com/godotengine/godot-proposals/assets/605817/8e7a7221-04ca-487d-b74f-337319daaa6c)

And this scene saves as follows:
```ini
[gd_scene load_steps=2 format=3 uid="uid://bwyd5lt8xj1h0"]

[ext_resource type="PackedScene" uid="uid://cfgv7xik5c3u5" path="res://main_console.tscn" id="1_qpby4"]

[node name="main" type="Control"]
layout_mode = 3

[node name="MainConsole" parent="." instance=ExtResource("1_qpby4")]
layout_mode = 1

[node name="ActiveArea" type="Control" parent="MainConsole" index="0"]
layout_mode = 2
```

Known issues that will need addressing:
- I haven't added a C# equivalent to `@export_default_child_path`, largely because I haven't compiled Godot with C# support yet. Presumably that will need to be added.
- If the subscene has Editable Children active at the time of save, children of the "default_child_path" node will be saved with their actual physical path, rather than "as children of the subscene root". That's probably the proper behavior, but I'd like other eyes on that.
- I haven't looked into what happens when you load a scene with an _editable_ subscene that has a default child path, if that subscene's _actual_ direct children are mentioned by name in the .tscn file. I think it'd probably work just fine but I haven't confirmed that.
- My intent is to make the "child path" property read-only in the inspector _unless_ you're editing that actual scene. Overriding the "child path" property as part of an instantiation has no effect, and this is intentional; the node mappings from number 4 above are created before the containing scene has a chance to set properties.
- I don't know what happens if a scene is loaded as placeholder, on account of the editor _doesn't_ load scenes as placeholder. Not sure how/where to test that.
- There are no checks to ensure that the `NodePath` actually points to a descendant of the subscene root, only that the `NodePath` points to _something_. You could probably get yourself in trouble if you manually specified that path with a `../` in front, but it'd require you to do so intentionally, or that you open a maliciously-crafted Godot scene. Is that a thing we're worried about? It doesn't seem like it probably is.

**Comment by theraot** [comment]
See also: https://github.com/godotengine/godot-proposals/issues/5475

**Comment by SlugFiller** [comment]
Looking at this again in detail, a couple of modifications to the approach come to mind:
1. There's no need to annotate as export, and export a `NodePath` property. This is redundant. Just specify the node path itself as an annotation parameter. This will remove the need to have and save an extra property. Annotations can already have parameters, and also apply to classes, rather than properties or methods. For example: `@icon`
2. Likewise, I'm not 100% sure, but saving the existence of the annotation in the tscn may be redundant as well. When instantiating the scene, any script on it is loaded, so as to run `_init`. This happens before instantiating any children, certainly before adding them as children. It is at this point that the annotation can be read from the script object directly. After that, the children can be added to the correct path.

Especially if the intention is for the path to be non-editable, putting it in the script instead of in an editable property just makes more sense.

**Comment by dmchurch** [comment]
> Looking at this again in detail, a couple of modifications to the approach come to mind:
> 
>     1. There's no need to annotate as export, and export a `NodePath` property. This is redundant. Just specify the node path itself as an annotation parameter. This will remove the need to have and save an extra property. Annotations can already have parameters, and also apply to classes, rather than properties or methods. For example: `@icon`

I'd considered that approach - the main reason I went with the annotated property was so that I could select the target path in the inspector. Easy enough to change to an annotation-with-argument on the script itself. The only downside of this is that it'd be a bit harder to make variants of the same scene with the "default child path" in a different place, but I think that use case may be way over on the edge.

>     2. Likewise, I'm not 100% sure, but saving the existence of the annotation in the tscn may be redundant as well. When instantiating the scene, any script on it is loaded, so as to run `_init`. This happens before instantiating any children, certainly before adding them as children. It is at this point that the annotation can be read from the script object directly. After that, the children can be added to the correct path.

I originally didn't have the extra annotation in the tscn (and, for that matter, I didn't have the extra field in `SceneState` either). I wasn't sure about if it would work under all scenarios, though, since scenes can be loaded as placeholder and so can scripts? Or so it seemed when I was looking at the code.

In the end I went with the tscn annotation because nothing else in the scene instantiation code pokes at script internals, and I figure that scene instantiation is one of those places that's critical enough that you don't want to be spending a whole bunch of extra cycles digging data out of script code.

It's also worth noting that having it explicit in the tscn means that it can be applied in ways other than "annotation in a script attached to the root element" - perhaps using the "slots" idea from #5475 or adding a class via GDExtension.

**Comment by SlugFiller** [comment]
> I'd considered that approach - the main reason I went with the annotated property was so that I could select the target path in the inspector.

It's possible to drag and drop nodes from the scene tree straight into the script window (which inserts a `NodePath` string), for roughly this sort of use case.

As for making variants of the scene, making an inherited script is as easy as, if not easier than, making an inherited scene.

**Comment by dmchurch** [comment]
> It's possible to drag and drop nodes from the scene tree straight into the script window (which inserts a `NodePath` string), for roughly this sort of use case.

Oh neat! I'd never tried that. Definitely worth making it a class-level annotation, I'll do that when I have a moment

**Comment by dmchurch** [comment]
@SlugFiller, I just want to check in with you - I'm working from the example of the `@icon` annotation, like you recommended, and this is turning out to be a more significant change than I expected: there isn't a generic "class metadata" API between the gdscript parser and the rest of the system, the way there's a generic "property export metadata" API. So, to pass the default child path from the gdscript out into the rest of Godot, I have to add a new abstract method to the `Script` class in `script_language.h` as well as implementations in the `GDScript`, `CSharpScript`, and `ScriptExtension` classes.

Both `@tool` and `@icon` use this codepath, so I know I'm not looking in totally the wrong place, but I worry about API explosion if more class-level annotations get added in the future - should I be looking into making some kind of generic class-level metadata channel between the script engines and Godot, or should I just add `get_default_child_path()` as another one-off to the Script API?

**Comment by dmchurch** [comment]
I actually just thought of a reason to keep this as a property declaration: putting it in a class-level annotation forces you to hard-code node paths into your script, which makes it harder to, say, use the same gdscript for multiple different scenes. I think this really needs to be editable on a per-scene basis, not just per-script.

I could imagine making it a special annotation that creates an invisible property that gets special treatment in the Inspector, but... why bother? That seems like it'd be a lot of work for very little benefit.

**Comment by SlugFiller** [comment]
> Both `@tool` and `@icon` use this codepath, so I know I'm not looking in totally the wrong place, but I worry about API explosion if more class-level annotations get added in the future - should I be looking into making some kind of generic class-level metadata channel between the script engines and Godot, or should I just add `get_default_child_path()` as another one-off to the Script API?

You should just add another one. Adding methods to an API when adding functionality is perfectly normal. Just consider how many methods are regularly being added to something like `RenderingServer`. If it's an abstract class then the implication is that this method has to be implemented by all subclasses, which means a GDExtensions and C# implementations become a must (Although you can put in stubs initially). Which is a good thing. You don't want such things to be "optional". You want it to be available to all interfaces.

**Comment by dmchurch** [comment]
Oh, okay! I'll just continue on the path I'm going, then 😄

**Comment by kitbdev** [comment]
I don't think this should even be in GDScript. It is entirely about the scene and has nothing to do with the script on the root node at all, and it doesn't affect the script functionality in any way. It is also useless or misleading if used in other scripts, and other languages would need to support it as well.
What it does it changes the parenting target in scene tree dock. A node in the scene tree needs to be marked as the default location for new children, and it needs to be saved with the scene. Having a script shouldn't be necessary at all.
Basically, I think #5475 is more clear and has the better workflow UX for this sort of feature.

**Comment by dmchurch** [comment]
> Basically, I think #5475 is more clear and has the better workflow UX for this sort of feature.

I'm fine with redirecting my efforts on this. @SlugFiller, what do you think? Should I turn this feature into a piece of metadata that gets saved with the scene and is activated by, say, a right-click interaction like enabling Editable Children? Take the script interactions out of it entirely?

I'll keep the existing styling I've added, where the default child node gets an "accent color" highlight and an explanatory message, like the one for the Unique Names icon - I think that works well for a "this node is significant in a way that none of the other nodes are" feature.

![image](https://github.com/godotengine/godot-proposals/assets/605817/265585a0-ed16-4aea-bf88-36912511916b)

**Comment by dmchurch** [comment]
The more I think about this, the more I think @kitbdev is 100% right about this not belonging in GDscript. Making this a property of the `Node` itself actually saves us from a number of really unfortunate UX implications, first and foremost "I can't rename my default child parent node without it losing its status".

It is still going to be a property unique within the context of a given Scene, however, just like Unique Names are, so I'll probably use those as a model for implementation.

**Comment by SlugFiller** [comment]
Yeah, the "This has nothing to do with GDScript" argument makes perfect sense if you think about it for more than 5 minutes. And since the implementation adds data to the tscn anyway, there's no reason why the source of that data wouldn't be in the `PackedScene` object itself.

Using unique nodes as a reference is indeed a good idea.

**Comment by dmchurch** [comment]
Interface for setting the default scene child parent:

![image](https://github.com/godotengine/godot-proposals/assets/605817/152d7506-293b-4847-a7b1-4d1954d38132)

which turns the node name blue and adds an icon:

![image](https://github.com/godotengine/godot-proposals/assets/605817/ef68e955-8635-4963-89e3-9419c69175d4)

Stores in the .tscn as a property on the scene root node:

```ini
[node name="MainConsole" type="VBoxContainer" node_paths=PackedStringArray("default_child_path", "clock", "transmission_state")]
default_child_path = NodePath("MainArea/CenterArea")
anchors_preset = 15
anchor_right = 1.0
anchor_bottom = 1.0
grow_horizontal = 2
grow_vertical = 2
theme = ExtResource("1_n65jp")
script = SubResource("GDScript_xm4y2")
clock = NodePath("HeaderBG/Header/Clock")
transmission_state = NodePath("HeaderBG/Header/TransmissionState")
```

which is also how it is stored internally, as a pointer on the scene node - that way it will track with the target object if you move it around the tree, if you rename it, etc; but since the property doesn't actually live on the child object, it won't get confused about copy/duplicate or anything like that.

Also has new GDScript-accessible API in `node.get_default_child_parent()` and `node.set_default_child_parent(n)`, which I'm sure the CI will yell at me for not adding to the docs :smile:

**Comment by Yahkub-R** [comment]
@dmchurch incredible stuff. I completely agree this is a better way of handling it over script based.  

How does this behave when you instantiate a scene? I would assume the children you add are still visible as direct children of the scene, but they visually appear in the default child parent? Something akin to this?  Where ContentContainer is the default child parent. 
![276730009-d71fb31a-9e95-4dc9-8472-3c57a2b9927e](https://github.com/godotengine/godot-proposals/assets/62478788/0806b443-c46e-4751-8f35-4d33d88c9fd8)

![276731888-002dec37-3415-4cc3-874e-fb7b293aea3d](https://github.com/godotengine/godot-proposals/assets/62478788/b5afbc24-dd95-46ff-8424-64ac61ef4177)

And how does it handle when you check editable children. Would hope it would show like this?
![276730583-20149c5d-ad21-487f-9505-104640bf7499](https://github.com/godotengine/godot-proposals/assets/62478788/75c5b82d-0535-4b63-8f74-d56bae168fac)

Really can't wait to use this as I was trying to tackle the same issue. Seems critical for any sort of consistent UI design.

**Comment by dmchurch** [comment]
Yes, that's exactly how it functions. I'll try and throw together an animation to showcase the behavior when I have a chance, but you've got the idea. I still have a few edge cases I want to exercise to make sure they don't cause havoc, but the following should all hold:

1. Create scene A with hierarchy A > B > C, set C as the DCP, save A
2. Create scene X, instantiate A, add node Y as child of A.
   - Hierarchy in scene tree dock: X > A > Y
   - Hierarchy saved to tscn: X > A > Y
   - actual instantiated `Node` hierarchy: X > A > B > C > Y
3. Enable "editable children" on X > A
   - Dock: X > A > {B} > {C} > Y (where the {} is in yellow)
   - tscn: X > A > [B] > [C] > Y (where [] are omitted in the source if unchanged, but Y is still called a child of C rather than A)
   - actual `Node`s: X > A > B > C > Y (unchanged)
4. Disable EC on X > A and save; returns to state 2.
5. Close X, add D as a child of C and set it as the DCP, save A with new hierarchy A > B > C > D
6. Reopen X.
   - Dock: X > A > Y
   - tscn: X > A > Y (obviously)
   - `Node`: X > A > B > C > D > Y

**Comment by dmchurch** [comment]
Adding a dashed line and tooltip to make it clearer when the Scene Tree is showing a discontiguous relationship:

![image](https://github.com/godotengine/godot-proposals/assets/605817/b729a6cd-238f-4c7b-b19a-b7aa67a1e17b)

(new API on `TreeItem`, property `indirect_child` sets whether the child in question should be rendered with a dashed line connected to its parent rather than a solid line)

**Comment by RandomShaper** [comment]
After some thinking, I'd say that this sort of magic relocation feature (_a la_ content projection in web frameworks) is a bit against the philosophy of Godot, by which everything is apparent. Now, even if we disregard such kind of argument, I think adding this raises the complexity of an already complex area (with consequences such as more difficult maintenance, bugs due to new edge cases, etc.), when maybe a different, more right-to-the-point approach could be taken.

> [...] but then I couldn't switch out my console or make changes to it without having to fix every place I've already instantiated it.
The main problem here is this, isn't it? The editable children feature looks like the way to go, but once you need to change your scene tree, you'll break each and single case of it being instantiated with children added. I wonder if it would be better to add an editor tool that tracks moves/renames of nodes containing children in derived scenes and provides the user with a one-click 'update all' button.

It could be something like this:

Assume you already have derivatives of `main_console.tscn` with nodes hanging from `CenterArea`. You load `main_console.tscn` in the editor, delete/move/rename `ActiveArea` and hit _Save_. The editor checks all the derivated scenes and finds that some would have stray nodes as a result. A dialog is shown warning you about the fact, with the following options:
- Abort save.
- Update all derivated scenes (relocate nodes). You'll pick the new parents for each moved/deleted/renamed node. The editor will try to guess the new correct ones (at least, if just moved, it can guess by name).

**Comment by theraot** [comment]
See also:
- https://github.com/godotengine/godot-proposals/issues/7053
- https://github.com/godotengine/godot-proposals/issues/7425

**Comment by aXu-AP** [comment]
I had an idea that is somewhat similiar to @raulsntos , altough it doesn't introduce a new Node. Allow to expose some nodes upon instancing. The exposed nodes have to have unique % names, so moving them around won't break anything.
![kuva](https://github.com/godotengine/godot-proposals/assets/1621768/e8a4bcf0-4474-493c-8468-2ee0c72a9fb4)

This would solve also an another problem I had a few times: if you use nodes as components (health, movement, ai etc.), I want sometimes change the parameters of these components on some instances. But to change their properties, I have to enable editable children, which comes it's own set of problems. With this system, I could just set components exposed.

**Comment by Yahkub-R** [comment]
@RandomShaper while i do generally agree that it adds complexity I think for the sake of user experience its worth it. In order to have the same functionality I would need to clutter my hierarchy with the 'editable children' enabled for my window 
node. vs simply instantiating my window node and adding a child.

@raulsntos, @aXu-AP I think this is a great approach that adds to what @dmchurch has been looking at. I would generally lean to @aXu-AP's approach as it doesn't add new nodes and is far more flexible than a single default child parent.  It also fixes the issue of having to link 2 nodes together to achieve the same effect.  

But, I do see a benefit in creating a "Exposed Control Node" or something of the sort similar to @raulsntos 's idea for the sake of clarity and the potential functionality of manipulating it via code etc.
